### PR TITLE
Add the ability to generate Ion struct based on the constraints in Io…

### PR DESCRIPTION
**Pull request by linlin-s**
Guide for reviewing:

GeneratorOption.java:
Add option to process Ion shcema file

ReadGeneralConstraints.java:
After reading the Ion schema file, go through the whole file and find the general constraints, like 'type', 'annotations', and other constraints relevant to the current data type. Then according to the value of 'type' to decide which data generator would be called, and the constraints information will also be passed to these methods.

WriteRandomIonValues.java:
Add method which can write requested size and format Ion struct. 
Add methods which can parse the constraints, all available constraints are from this set (timestamp_precision | codepoint_length | occurs | container_length | annotations).

textStructs.isl: this is the ion-schema file

Note: For the unit test, I'm thinking about using ion schema kotlin to check the validation of generated data. Unit test will be added in the next PR.

### CODE REVIEW

File: src/com/amazon/ion/benchmark/GeneratorOptions.java [hunk 1]
```diff
@@ -28,15 +28,18 @@ public static void executeGenerator(Map<String, Object> optionsMap) throws Excep
         } else {
             timestampTemplate = optionsMap.get("--timestamps-template").toString();
         }
-
+        if (optionsMap.get("--input-ion-schema") != null) {
+            String inputFilePath = optionsMap.get("--input-ion-schema").toString();
+            ReadGeneralConstraints.readIonSchemaAndGenerate(size, inputFilePath, format, path);
+        }
         if (optionsMap.get("--data-type") != null) {
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/GeneratorOptions.java]: Should this be `else if`? Is it allowed for the user to specify both `--input-ion-schema` and `--data-type`?

File: src/com/amazon/ion/benchmark/GeneratorOptions.java [hunk 2]
```diff
@@ -28,15 +28,18 @@ public static void executeGenerator(Map<String, Object> optionsMap) throws Excep
...
             IonType type = IonType.valueOf(optionsMap.get("--data-type").toString().toUpperCase());
             switch (type) {
                 case TIMESTAMP:
                     WriteRandomIonValues.writeRandomTimestamps(size, type, path, timestampTemplate, format);
                     break;
                 case STRING:
-                    WriteRandomIonValues.writeRandomStrings(size, type, path, range, format);
+                    WriteRandomIonValues.writeRandomStrings(size, type, path, range, format,20);
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/GeneratorOptions.java]: Why 20? (Also, add a space between the comma and the 2.)

File: src/com/amazon/ion/benchmark/GeneratorOptionsValidator.java [hunk 3]
```diff
@@ -16,10 +16,10 @@
...
-    final public static Set<String> invalidForDecimal = Collections.unmodifiableSet(new HashSet<>(Arrays.asList("-N", "--text-code-point-range", "-M", "--timestamps-template")));
-
+    final public static Set<String> invalidForTimestamp = Collections.unmodifiableSet(new HashSet<>(Arrays.asList("-N", "--text-code-point-range", "-E", "--decimal-exponent-range <exp_range>", "-C", "--decimal-coefficient-digit-range", "-Q", "--input-ion-schema")));
+    final public static Set<String> invalidForString = Collections.unmodifiableSet(new HashSet<>(Arrays.asList("-M", "--timestamps-template", "-E", "--decimal-exponent-range <exp_range>", "-C", "--decimal-coefficient-digit-range", "-Q", "--input-ion-schema")));
+    final public static Set<String> invalidForDecimal = Collections.unmodifiableSet(new HashSet<>(Arrays.asList("-N", "--text-code-point-range", "-M", "--timestamps-template", "-Q", "--input-ion-schema")));
+    final public static Set<String> invalidForIonSchema = Collections.unmodifiableSet(new HashSet<>(Arrays.asList("-N", "--text-code-point-range", "-E", "--decimal-exponent-range <exp_range>", "-C", "--decimal-coefficient-digit-range","-M", "--timestamps-template")));
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/GeneratorOptionsValidator.java]: What about `--data-type`?

File: src/com/amazon/ion/benchmark/Main.java [hunk 4]
```diff
@@ -260,7 +260,8 @@
             + "The string will be generated by choosing a random unicode code points that are within the given range. When the option is "
             + "not provided, this range will be set to the default value. [default: [0,1114111]]\n"
 
-        + "  -Q --input-ion-Schema <file_path>\n"
+        + "  -Q --input-ion-schema <file_path>      This option will specify the path of ion schema file which contains all constraints that the "
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/Main.java]: Capitalize Ion Schema

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * Get general constraints of Ion schema and call the relevant generator method based on the type.
+     * @param size is the size of the output file.
+     * @param path is the path of the ion schema file.
+     * @param format is the format of the generated file, select from set (ion_text | ion_binary).
+     * @param outputFile is the path of the generated file.
+     * @throws Exception if errors occur when reading and writing data.
+     */
+    public static void readIonSchemaAndGenerate(int size, String path, String format, String outputFile) throws Exception {
+        IonReader reader = IonReaderBuilder.standard().build(new BufferedInputStream(new FileInputStream(path)));
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: The reader needs to be closed when reading is complete.

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        IonSystem system = IonSystemBuilder.standard().build();
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: This system can be a constant (i.e. declare it as a `private static final` variable in this class.

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        IonLoader loader = system.newLoader();
+        IonDatagram schemaValue = loader.load(reader);
+
+        for (int i = 0; i < schemaValue.size(); i++) {
+            // Assume there's only one constraint between schema_header and schema_footer, if more constraints added, here is the point where developers should start.
+            if (schemaValue.get(i).getType().equals(IonType.STRUCT) && schemaValue.get(i).getTypeAnnotations()[0].equals("type")) {
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: Before the `if`, set `schemaValue.get(i)` to a variable so it can be reused without having to get by index.

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                IonStruct constraintStruct = (IonStruct) schemaValue.get(i);
+                //get general constraints:
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: Line comments should have  a space between the second / and the first letter (which should be capitalized).

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 9]
```diff
@@ -84,7 +84,7 @@ private static IonWriter formatWriter(String format, File file) throws Exception
         List<Integer> result = new ArrayList<>();
         while (reader.next() != null) {
             if (reader.getType() != IonType.INT) throw new IllegalStateException("Please put integer elements inside of the list");
-            int value = reader.intValue();
+            Integer value = reader.intValue();
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: Why was this changed?

File: textStructs.isl [hunk 10]
```diff
@@ -0,0 +1,25 @@
+$ion_schema_1_0
```
Comment (tgregg) [inline textStructs.isl]: This file will only be needed for tests, right? It should go under the `tst` directory.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 11]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+        }
+        return precision;
+    }
+
+    /**
+     * Return the value of annotations, this method is available when the annotations are required, if there are more constraints relative to the annotations, more logic needed to be added here.
+     * @param value is the Ion struct which contain the current constraint field.
+     * @return an Ion List contains all annotations.
+     */
+    public static IonList getAnnotation(IonStruct value) {
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: I recommend putting all the constants and any utility functions related to Ion Schema (like this one) into a separate class (named something like `IonSchemaUtilities`). That way all of the stuff specific to Ion Schema is centrally located.

Basically, any time you are navigating the structure of an IonValue that represents an Ion Schema, put that utility function in the new location, and make any hard-coded Ion Schema field names (e.g. "required", "codepoint_length") constants.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 12]
```diff
@@ -562,21 +503,26 @@ private static void writeDataToFile(IonType type, IonWriter writer, Random rando
...
+                WriteRandomIonValues.constructInt(random, writer);
                 break;
             case TIMESTAMP:
                 Object timestampTemplate = objects[4];
                 Timestamp timestamp = WriteRandomIonValues.constructTimestamp(timestampTemplate);
                 writer.writeTimestamp(timestamp);
                 break;
+            case STRUCT:
+                IonStruct fields = (IonStruct) objects[4];
+                IonList annotations = (IonList) objects[5];
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: I know this wasn't changed in this PR (and I probably should have caught it before), but we shouldn't be using `Object... objects` in method signatures to allow the method to be provided with any arguments. It leads to too many lines of code like this, which are hard to change if the implicit spec for what's allowed in the `objects` array is changed. I recommend creating named parameters for everything in the method signatures, even if some of the parameters are unused by some of the types (this change will affect several of the methods in this class).

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 13]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+    public static void constructIonStruct(IonStruct fields, IonList annotations, IonWriter writer) throws IOException {
+        Random random = new Random();
+        try (IonReader reader = IonReaderBuilder.standard().build(fields)) {
+            reader.next();
+            reader.stepIn();
+            int index = random.nextInt(annotations.size());
+            writer.addTypeAnnotation(String.valueOf(annotations.get(index)));
+            writer.stepIn(IonType.STRUCT);
+            while (reader.next() != null) {
+                IonSystem system = IonSystemBuilder.standard().build();
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: Reuse this as a constant.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 14]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+            reader.next();
+            reader.stepIn();
+            while (reader.next() != null) {
+                if (reader.getFieldName().equals(keyWord)) {
+                    IonType type = reader.getType();
+                    switch (type) {
+                        case INT:
+                            result = reader.intValue();
+                            break;
+                        case SYMBOL:
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: Does the Ion Schema spec specify that this *must* be a symbol, or may it also be a string?

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 15]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+                            if (reader.stringValue().equals("required")) {
+                                result = 1;
+                            } else {
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: I recommend making the code more explicit by making this an `else if` that checks for "optional". Then your `else` branch can throw an `IllegalArgumentException` or something because it would mean that the schema contained something not supported.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 16]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+                    continue;
+                }
+                writer.setFieldName(fieldName);
+                IonType type = null;
+                type = IonType.valueOf(value.get("type").toString().toUpperCase());
+                switch (type) {
+                    // if more types of Ion data are available, the logic should be added below.
+                    case STRING:
+                        String codepointRange = "[0, 1114111]";
+                        List<Integer> range = parseRange(codepointRange);
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: This range looks like it should be a constant.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java [hunk 17]
```diff
@@ -698,4 +642,158 @@ public static void constructLobs(Random random, IonType type, IonWriter writer)
...
+                        int codePointsLengthBound = 20;
+                        if (value.get("codepoint_length") != null) {
+                            codePointsLengthBound = WriteRandomIonValues.parseConstraints(value, "codepoint_length");
+                        }
+                        String stringValue = WriteRandomIonValues.constructString(range, codePointsLengthBound);
+                        writer.writeString(stringValue);
+                        break;
+                    case TIMESTAMP:
+                        // call the function that extract the precision of the timestamp
+                        Timestamp.Precision[] precisions = Timestamp.Precision.values();
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: `Timestamp.Precision.values()` will allocate a new array each time it is called. Since it never changes, consider storing its result as a constant.

**Review by tgregg — COMMENTED** [review]
Good start. I can see how you're going to extend this to other types and constraints.

For the next PR, let's just create it directly in the `amzn` repo.

File: src/com/amazon/ion/benchmark/GeneratorOptions.java (hunk 1 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/GeneratorOptions.java:35]: > Should this be `else if`? Is it allowed for the user to specify both `--input-ion-schema` and `--data-type`?
It should be else if , and they're not allowed to be specified at the same time.

File: src/com/amazon/ion/benchmark/GeneratorOptions.java (hunk 2 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/GeneratorOptions.java]: > Why 20? (Also, add a space between the comma and the 2.)

It should be a random.nextInt(20), I will update the logic. The default value of length boundary in [existing code ](https://gist.github.com/tgregg/d557f1ec858a2461e23a370fda66c6dc) is 20 , so I just keep using this value

File: src/com/amazon/ion/benchmark/GeneratorOptionsValidator.java (hunk 3 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/GeneratorOptionsValidator.java]: > What about `--data-type`?

'--data-type' is added.

File: src/com/amazon/ion/benchmark/Main.java (hunk 4 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/Main.java]: > Capitalize Ion Schema

Change has been updated

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java (hunk 5 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: > The reader needs to be closed when reading is complete.

Change has been updated

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java (hunk 6 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: > This system can be a constant (i.e. declare it as a `private static final` variable in this class.

Changes have been updated

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java (hunk 7 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: > Before the `if`, set `schemaValue.get(i)` to a variable so it can be reused without having to get by index.

Change has been updated.

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java (hunk 8 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/ReadGeneralConstraints.java]: > Line comments should have a space between the second / and the first letter (which should be capitalized).

Changes have been updated

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 9 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > Why was this changed?

This was changed by mistake, this line has been updated to the original version

File: textStructs.isl (hunk 10 above)
Comment (linlin-s) [reply textStructs.isl]: > This file will only be needed for tests, right? It should go under the `tst` directory.

Yes, for test only

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 11 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java:791]: > I recommend putting all the constants and any utility functions related to Ion Schema (like this one) into a separate class (named something like `IonSchemaUtilities`). That way all of the stuff specific to Ion Schema is centrally located.
> 
> Basically, any time you are navigating the structure of an IonValue that represents an Ion Schema, put that utility function in the new location, and make any hard-coded Ion Schema field names (e.g. "required", "codepoint_length") constants.
All relevant variables and functions have been moved to a separated class, and all these hard-coded names have been assigned as constants.

Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > I recommend putting all the constants and any utility functions related to Ion Schema (like this one) into a separate class (named something like `IonSchemaUtilities`). That way all of the stuff specific to Ion Schema is centrally located.
> 
> Basically, any time you are navigating the structure of an IonValue that represents an Ion Schema, put that utility function in the new location, and make any hard-coded Ion Schema field names (e.g. "required", "codepoint_length") constants.

All relevant variables and functions have been moved to a separated class, and all these hard-coded names have been assigned as constants.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 12 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > I know this wasn't changed in this PR (and I probably should have caught it before), but we shouldn't be using `Object... objects` in method signatures to allow the method to be provided with any arguments. It leads to too many lines of code like this, which are hard to change if the implicit spec for what's allowed in the `objects` array is changed. I recommend creating named parameters for everything in the method signatures, even if some of the parameters are unused by some of the types (this change will affect several of the methods in this class).

Changes have been updated.

File: src/com/amazon/ion/benchmark/GeneratorOptions.java (hunk 1 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/GeneratorOptions.java]: > Should this be `else if`? Is it allowed for the user to specify both `--input-ion-schema` and `--data-type`?

 It should be else if , and they're not allowed to be specified at the same time.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 14 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > Does the Ion Schema spec specify that this _must_ be a symbol, or may it also be a string?

It must be a symbol

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 13 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > Reuse this as a constant.

Change has been updated

Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java:678]: > Reuse this as a constant.

Change has been updated

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 14 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java:739]: > Does the Ion Schema spec specify that this _must_ be a symbol, or may it also be a string?

It must be a symbol

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 15 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > I recommend making the code more explicit by making this an `else if` that checks for "optional". Then your `else` branch can throw an `IllegalArgumentException` or something because it would mean that the schema contained something not supported.

Sounds great! Change has been updated.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 16 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > This range looks like it should be a constant.

This part is updated as well.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java (hunk 17 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > `Timestamp.Precision.values()` will allocate a new array each time it is called. Since it never changes, consider storing its result as a constant.

Set all precisions as a constant, changes have been updated.

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java [hunk 18]
```diff
@@ -7,70 +7,73 @@
...
 import java.io.FileInputStream;
 
 /**
- * Parse Ion schema file and get the general constraints in the file then pass the constraints to the Ion data generator.
+ * Parse Ion Schema file and get the general constraints in the file then pass the constraints to the Ion data generator.
  */
 public class ReadGeneralConstraints {
+    private static final IonSystem system = IonSystemBuilder.standard().build();
+    private static final IonLoader loader = system.newLoader();
+    private static final String defaultRange = "[0, 1114111]";
```
Comment (tgregg) [inline src/com/amazon/ion/benchmark/ReadGeneralConstraints.java:23]: Minor: Java style recommends all-caps snake case for constants, e.g. `SYSTEM`, `DEFAULT_RANGE`.

**Review by tgregg — APPROVED** [review]
Looks good!

File: src/com/amazon/ion/benchmark/ReadGeneralConstraints.java (hunk 18 above)
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/ReadGeneralConstraints.java:23]: > Minor: Java style recommends all-caps snake case for constants, e.g. `SYSTEM`, `DEFAULT_RANGE`.

Got it . I'll update these changes in the next PR

**Review by linlin-s — COMMENTED** [review]
I'll merge this PR first and the minor change will be updated in the next PR
